src/wrappers.js

```diff
@@ -100,3 +102,225 @@ wrappers[".google.protobuf.Any"] = {
+            struct[names[i]] = googleProtobufValueToObject(message.structValue["fields"][names[i]]);
+        }
+        return struct;
+    }
+    return undefined;
+}
+
+// custom wrapper for google.protobuf.Value
+wrappers[".google.protobuf.Value"] = {
+    fromObject: function(object) {
+        // heuristic: if an object looks like a regular representation of google.protobuf.Value,
+        // with all those stringValues, etc., just accept it as is for compatibility.
+        if (typeof object === "object" && object) {
+            // something that has just one property called stringValue, listValue, etc.,
+            // and possibly a property called kind, is likely an object we don't want to touch
+            var names = Object.keys(object);
+            if (names.length === 1 && names.match(/^(?:null|number|string|list|struct)Value$/) ||
+                names.length === 2 &&
+                    names.every(function(name) {
+                        return name.match(/^(?:kind|(?:null|number|string|list|struct)Value)$/); })
+            ) {
+                return this.fromObject(object);
+            }
+        }
+
+        // otherwise, it's a JSON representation as described in google/protobuf/struct.proto
+        var self = this;
+        var message = googleProtobufValueFromObject(object, function(obj) { return self.create(obj); });
+        if (typeof message !== "undefined") {
+            return message;
+        }
+
+        // fallback to the normal .fromObject if decoding failed
+        return this.fromObject(object);
+    },
+
+    toObject: function(message, options) {
+        // decode value if requested
+        // In the next major version we will get rid of "options.values".
+        if (options && options.json && options.values) {
+            var object = googleProtobufValueToObject(message);
+            if (typeof object !== "undefined") {
+                return object;
+            }
+        }
+
+        return this.toObject(message, options);
+    }
+};
+
+// custom wrapper for google.protobuf.Struct
+wrappers[".google.protobuf.Struct"] = {
+    fromObject: function(object) {
+        if (typeof object === "object" && object) {
+            var names = Object.keys(object),
+                i = 0,
+                fields = {};
+
+            // heuristic: if an object looks like a regular representation of google.protobuf.Struct,
+            // with just one field called `fields`, just accept it as is for compatibility.
+            if (names.length === 1 && names[0] === "fields") {
+                return this.fromObject(object);
+            }
+
+            // decoding Struct requires Value; make a fake Root to get it
+            var root = new Root().addJSON(common["google/protobuf/struct.proto"].nested).resolveAll();
+            var Value = root.lookup("google.protobuf.Value");
+
+            for (; i < names.length; ++i) {
+                fields[names[i]] = googleProtobufValueFromObject(
```

Comment:
Why not just make this `fields[names[i]] = Value.fromObject(object[names[i]])`? That should be properly delegated to the `fromObject` wrapper here.

Reply:
Done, thank you! (changed both here and in `toObject`)